src/Native/Unix/System.Native/pal_io.c

```diff
@@ -441,6 +441,25 @@ int32_t SystemNative_ReadDirR(DIR* dir, uint8_t* buffer, int32_t bufferSize, Dir
     struct dirent* result = NULL;
     int error = readdir_r(dir, entry, &result);
 
+#ifdef _AIX
+    // AIX returns 0 on success, but bizarrely, it returns 9 for both error and
+    // end-of-directory - in the latter case, it sets result to NULL, and in the
+    // former, sets errno. (I'm not making this up either, see readdir_r(3).)
+
+    // 9 returned, do we have a null result for end of directory?
+    if (error == 9 && result == NULL)
+    {
+        memset(outputEntry, 0, sizeof(*outputEntry)); // managed out param must be initialized
+        return -1;         // shim convention for end-of-stream
+    }
+    // result non-null, so must be an error
+    else if (error == 9)
+    {
+        assert(error > 0);
```

Comment:
if it's 9 it's definitely going to be > 0

Reply:
Was this meant to be errno rather than error?